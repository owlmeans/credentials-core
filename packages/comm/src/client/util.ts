/**
 *  Copyright 2023 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import {
  CredentialWrapper, Credential, ExtensionRegistry, EXTENSION_TRIGGER_INCOMMING_DOC_RECEIVED, RegistryItem,
  WalletWrapper, Presentation
} from "@owlmeans/vc-core"
import { IncommigDocumentWithConn, IncommingWrapperMeta } from "../types"


export const triggerIncommingDocView = async <
  Subject extends {} = {},
  Type extends RegistryItem<Subject> = Presentation<Credential<Subject>>,
  Meta extends IncommingWrapperMeta = IncommingWrapperMeta
>(
  registry: ExtensionRegistry, wallet: WalletWrapper, wrapper: CredentialWrapper<Subject, Type, Meta>
) => {
  await registry.triggerEvent<IncommigDocumentWithConn>(
    wallet, EXTENSION_TRIGGER_INCOMMING_DOC_RECEIVED, {
    credential: wrapper.credential as unknown as Presentation, statusHandler: { successful: false },
    conn: (wrapper.meta as IncommingWrapperMeta).conn,
    cleanUp: () => undefined
  })
}