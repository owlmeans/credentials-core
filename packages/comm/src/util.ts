/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { generateKeyPairFromSeed } from '@stablelib/x25519'
import {
  CryptoKey, DIDDocument, DIDHelper, DIDPURPOSE_AGREEMENT, ExtensionRegistry, KEYCHAIN_ERROR_NO_KEY,
  KeyPairToCryptoKeyOptions, Credential, Presentation, WalletWrapper, ERROR_NO_IDENTITY
} from "@owlmeans/regov-ssi-core"
import {
  CommKey, COMM_DID_AGREEMENT_KEY_DEFAULT, connectionFieldList, DIDCommConnectMeta,
  DIDCommHelper, DIDCommListner, ERROR_COMM_CANT_SEND, ERROR_COMM_DID_NOKEY, ERROR_COMM_NO_RECIPIENT,
  ERROR_COMM_WS_UNKNOWN, EVENT_INIT_CONNECTION, InitCommEventParams
} from "./types"
import { JWE } from 'did-jwt'


export const cryptoKeyToCommKey = async (
  key: CryptoKey, options?: KeyPairToCryptoKeyOptions
): Promise<CommKey> => {
  if (!key.pk) {
    throw Error(KEYCHAIN_ERROR_NO_KEY)
  }
  const pk = Buffer.from(key.pk, 'utf8').subarray(0, 32)
  const pubKey = generateKeyPairFromSeed(pk).publicKey

  return {
    id: options?.id || key.id || '_noid_',
    pk, pubKey
  }
}

export const filterConnectionFields = (connection: DIDCommConnectMeta): DIDCommConnectMeta => {
  return Object.fromEntries(
    connectionFieldList.map(field => [field, connection[field]])
  ) as unknown as DIDCommConnectMeta
}

export const invertConnection = (connection: DIDCommConnectMeta, channel?: string): DIDCommConnectMeta => {
  const newConnection = filterConnectionFields(JSON.parse(JSON.stringify(connection)))

  if (!newConnection.recipient) {
    throw new Error(ERROR_COMM_NO_RECIPIENT)
  }

  return {
    ...newConnection,
    recipientId: newConnection.sender.id,
    recipient: newConnection.sender,
    sender: newConnection.recipient,
    channel: channel || newConnection.channel
  }
}

export const parseJWE = (data: string) => {
  try {
    return JSON.parse(data) as JWE
  } catch (e) {
  }

  return undefined
}

export const didDocToCommKeyBuilder = (helper: DIDHelper) =>
  async (did: DIDDocument, keyId: string = COMM_DID_AGREEMENT_KEY_DEFAULT): Promise<Partial<CommKey>> => {
    const agreement = helper.expandVerificationMethod(did, DIDPURPOSE_AGREEMENT, keyId)
    if (!agreement.publicKeyBase58) {
      throw new Error(ERROR_COMM_DID_NOKEY)
    }

    return {
      id: did.id,
      pubKey: helper.getCrypto().base58().decode(agreement.publicKeyBase58)
    }
  }

export const getDIDCommUtils = (wallet: WalletWrapper) => {
  return {
    send: async (
      conn: DIDCommConnectMeta, doc?: Credential | Presentation,
      listener?: (helper: DIDCommHelper) => DIDCommListner
    ): Promise<DIDCommHelper> => {
      const ext = wallet.getExtensions()
      if (!ext) {
        throw ERROR_COMM_WS_UNKNOWN
      }

      return await new Promise(async (resolve) => {
        resolve(await new Promise(async (resolve, reject) => {
          let sentOnce = false
          await ext.triggerEvent<InitCommEventParams>(wallet, EVENT_INIT_CONNECTION, {
            statusHandle: { established: false },
            resolveConnection: async (helper) => {
              if (sentOnce) {
                return
              }
              sentOnce = true
              console.log('TRY CONNECTION', conn)
              if (!wallet.did.helper().isDIDDocument(conn.sender)) {
                throw ERROR_NO_IDENTITY
              }

              let toReturn: DIDCommListner | undefined = undefined
              if (listener) {
                toReturn = listener(helper)
                await helper.addListener(toReturn)
              }

              if (doc) {
                if (!await helper.send(doc, conn)) {
                  toReturn && await helper.removeListener(toReturn)
                  reject(ERROR_COMM_CANT_SEND)
                }
                toReturn && await helper.removeListener(toReturn)
              }

              resolve(helper)
            },
            rejectConnection: async (err) => {
              console.error(err)
              reject(err)
            }
          })
        }))
      })
    }
  }
}

