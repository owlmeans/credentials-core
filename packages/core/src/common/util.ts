/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { MaybeArray, Idish } from "./types"

export const normalizeValue = <Type>(value: MaybeArray<Type>): Type[] => {
  return Array.isArray(value)
    ? value
    : value ? [value] : []
}

export const simplifyValue = <Type>(value: MaybeArray<Type>) => {
  return Array.isArray(value) && value.length < 2
    ? value.length === 0 ? undefined : value[0]
    : value
}

export const singleValue = <Type>(value: MaybeArray<Type>): Type | undefined => {
  return Array.isArray(value)
    ? value.length > 0 ? value[0] : undefined
    : value
}

export const mapValue = async <Type, Res>(
  value: MaybeArray<Type>,
  callback: (value: Type) => Promise<Res>
): Promise<Res[]> => {
  return Promise.all(normalizeValue(value).map(callback))
}

export const addToValue = <Type>(value: MaybeArray<Type> | undefined, add: MaybeArray<Type>): MaybeArray<Type> => {
  if (value) {
    return [...normalizeValue(value), ...normalizeValue(add)]
  }

  return add
}

export const extractId = (id: Idish) => typeof id === 'string' ? id : id.id

export const getDeepValue = <Type, Subject extends {}>(subject: Subject, key: string, def?: Type): Type => {
  const parts = key.split('.')
  return parts.reduce((value, key) => {
    if (subject.hasOwnProperty(key)) {
      const tmp = (subject as any)[key]
      subject = tmp

      return tmp
    }

    return def
  }, def) as Type
}