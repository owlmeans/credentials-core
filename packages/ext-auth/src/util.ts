import { Credential, Presentation } from "@owlmeans/regov-ssi-core";
import { AuthSubject, REGOV_AUTH_RESPONSE_TYPE, REGOV_CREDENTIAL_TYPE_AUTH } from "./types";



export const getAuthFromPresentation = (presentation: Presentation) => {
  if (!presentation.type.includes(REGOV_AUTH_RESPONSE_TYPE)) {
    return undefined
  }

  return presentation.verifiableCredential.find(
    credential => credential.type.includes(REGOV_CREDENTIAL_TYPE_AUTH)
  )
}

export const getAuthSubject = (cred: Credential): AuthSubject => {
  return cred.credentialSubject as unknown as AuthSubject
}

export const pinValidation = {
  required: true,
  minLength: 4,
  pattern: /^\d+$/
}