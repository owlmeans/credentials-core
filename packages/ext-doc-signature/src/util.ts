/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { MaybeArray, normalizeValue } from "@owlmeans/regov-ssi-core"
import { Presentation, Credential } from "@owlmeans/regov-ssi-core"
import {
  BASIC_IDENTITY_TYPE, REGOV_SIGNATURE_REQUEST_TYPE, REGOV_SIGNATURE_RESPONSE_TYPE,
  REGOV_CREDENTIAL_TYPE_SIGNATURE
} from "./types"


export const getSignatureRequestFromPresentation = (presentation: Presentation) => {
  if (!presentation.type.includes(REGOV_SIGNATURE_REQUEST_TYPE)) {
    return undefined
  }

  return presentation.verifiableCredential.find(
    credential => credential.type.includes(REGOV_SIGNATURE_REQUEST_TYPE)
  )
}

export const getSignatureResponseFromPresentation = (presentation: Presentation) => {
  if (!presentation.type.includes(REGOV_SIGNATURE_RESPONSE_TYPE)) {
    return undefined
  }

  return presentation.verifiableCredential.find(
    credential => credential.type.includes(REGOV_CREDENTIAL_TYPE_SIGNATURE)
  )
}

export const getSignatureRequestOwner = (crednetial: Credential) => {
  if (!crednetial.type.includes(REGOV_SIGNATURE_REQUEST_TYPE)) {
    return undefined
  }

  return normalizeValue<Credential>(crednetial.evidence as MaybeArray<Credential>).find(
    evidence => evidence?.type.includes(BASIC_IDENTITY_TYPE)
  )
}