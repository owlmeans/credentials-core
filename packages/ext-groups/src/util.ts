/**
 *  Copyright 2023 OwlMeans, Inc
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { MaybeArray, normalizeValue } from "@owlmeans/vc-core"
import { Presentation, Credential } from "@owlmeans/vc-core"
import { DIDDocument } from "@owlmeans/vc-core"
import {
  OWLMEANS_CREDENTIAL_TYPE_GROUP, OWLMEANS_CREDENTIAL_TYPE_MEMBERSHIP, OWLMEANS_MEMBERSHIP_CLAIM_TYPE,
  BASIC_IDENTITY_TYPE, OWLMEANS_MEMBERSHIP_OFFER_TYPE
} from "./types"


export const getGroupFromMembershipClaimPresentation = (presentation: Presentation) => {
  if (!presentation.type.includes(OWLMEANS_MEMBERSHIP_CLAIM_TYPE)) {
    return undefined
  }

  const membershipClaim = presentation.verifiableCredential.find(
    credential => credential.type.includes(OWLMEANS_CREDENTIAL_TYPE_MEMBERSHIP)
  )

  return normalizeValue(membershipClaim?.evidence).find(
    evidence => evidence?.type.includes(OWLMEANS_CREDENTIAL_TYPE_GROUP)
  ) as Credential
}

export const getGroupOwnerIdentity = (crednetial: Credential) => {
  if (!crednetial.type.includes(OWLMEANS_CREDENTIAL_TYPE_GROUP)) {
    return undefined
  }

  return normalizeValue<Credential>(crednetial.evidence as MaybeArray<Credential>).find(
    evidence => evidence?.type.includes(BASIC_IDENTITY_TYPE)
      && evidence.id === (crednetial.issuer as unknown as DIDDocument).id
  )
}

export const getMembershipClaimHolder = (presentation: Presentation) => {
  const membership = presentation.verifiableCredential.find(
    credential => credential.type.includes(OWLMEANS_CREDENTIAL_TYPE_MEMBERSHIP)
  )

  return membership?.issuer as unknown as DIDDocument
}

export const getMembershipClaim = (presentation: Presentation) => {
  return presentation.verifiableCredential.find(
    credential => credential.type.includes(OWLMEANS_CREDENTIAL_TYPE_MEMBERSHIP)
  )
}

export const getGroupFromMembershipOfferPresentation = (presentation: Presentation) => {
  if (!presentation.type.includes(OWLMEANS_MEMBERSHIP_OFFER_TYPE)) {
    return undefined
  }

  const membershipClaim = presentation.verifiableCredential.find(
    credential => credential.type.includes(OWLMEANS_CREDENTIAL_TYPE_MEMBERSHIP)
  )

  return normalizeValue(membershipClaim?.evidence).find(
    evidence => evidence?.type.includes(OWLMEANS_CREDENTIAL_TYPE_GROUP)
  ) as Credential
}

export const getMembershipOffer = (presentation: Presentation) => {
  return presentation.verifiableCredential.find(
    credential => credential.type.includes(OWLMEANS_CREDENTIAL_TYPE_MEMBERSHIP)
  )
}