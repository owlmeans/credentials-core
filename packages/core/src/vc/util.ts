/**
 *  Copyright 2023 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */


import { MaybeArray } from "../common"
import { CredentialSchema, FullCredentialSchema, FullCrednetialEvidnce, UnsignedCredential } from "./types"
import { Presentation, Credential, Evidence } from "./types"


export const isPresentation = (obj: Object): obj is Presentation => {
  return obj.hasOwnProperty('verifiableCredential')
}

export const isCredential = (obj: Object): obj is Credential => {
  return obj.hasOwnProperty('credentialSubject')
}

export const extractSubject = <
  CredentialT extends UnsignedCredential = Credential
>(cred: CredentialT, idx?: number) => {
  type SubjectT = CredentialT extends UnsignedCredential<infer T>
    ? T extends Array<infer ST>
    ? ST : T
    : never

  const subject = Array.isArray(cred.credentialSubject)
    ? cred.credentialSubject[idx || 0]
    : cred.credentialSubject

  return subject as SubjectT
}

export const isFullEvidence = (obj: Evidence): obj is FullCrednetialEvidnce => {
  return obj.hasOwnProperty('credentialSubject')
}

export const isFullCredentialSchema = (obj: CredentialSchema): obj is FullCredentialSchema => {
  return obj.hasOwnProperty('credentialSubject')
}

export const getCompatibleSubject = <Type extends MaybeArray<{}>>(cred: UnsignedCredential): Type => {
  return cred.credentialSubject as unknown as Type
}