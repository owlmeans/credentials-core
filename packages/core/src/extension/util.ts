/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { normalizeValue } from "../common"
import { BasicCredentialType } from "../vc"
import {
  CredentialService, CredentialServiceBuilder, defaultBuildMethod, defaultClaimMethod,
  defaultOfferMethod, defaultRequestMethod, defaultRespondMethod, defaultSignMethod, defaultValidateMethod,
  Extension
} from "./ext"
import { CredentialDescription } from "./schema"


export const findAppropriateCredentialType = (
  ext: Extension, types: BasicCredentialType, defaultType: string
): string => {
  types = normalizeValue(types)

  return (Object.entries(ext.factories).map(([type]) => type)
    .find(type => types.includes(type)) || defaultType)
}

const methodMap = {
  'produceBuildMethod': 'build',
  'produceSignMethod': 'sign',
  'produceValidateMethod': 'validate',
  'produceClaimMethod': 'claim',
  'produceOfferMethod': 'offer',
  'produceRequestMethod': 'request',
  'produceRespondMethod': 'respond'
}

export const addFactoriesToExt = (ext: Extension, type: string, factories: CredentialServiceBuilder) => {
  const description = (ext.schema.credentials as { [key: string]: CredentialDescription })[type]
  ext.factories[type] = {
    ...{
      build: defaultBuildMethod(description),
      sign: defaultSignMethod(description),
      validate: defaultValidateMethod(description),
      claim: defaultClaimMethod(description),
      offer: defaultOfferMethod(description),
      request: defaultRequestMethod(description),
      respond: defaultRespondMethod(description)
    },
    ...Object.entries(factories).reduce((_facts, [method, builder]) => {
      if (ext.schema.credentials) {
        return {
          ..._facts,
          [(methodMap as any)[method]]: builder(description)
        }
      }
      return _facts
    }, {} as CredentialService)
  }
}