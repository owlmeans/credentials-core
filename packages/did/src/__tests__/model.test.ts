require('dotenv').config()

import { 
  buildDidHelper,
  DIDDocument, DIDPURPOSE_ASSERTION, DIDPURPOSE_VERIFICATION,
  buildDidRegistryWarpper, VERIFICATION_KEY_CONTROLLER
} from "../index"
import { nodeCryptoHelper } from "@owlmeans/regov-ssi-common"

import util from 'util'
util.inspect.defaultOptions.depth = 8


const testContext: {
  didDoc?: DIDDocument
  holderDoc?: DIDDocument
} = {}

describe('DID Helper', () => {
  const didHelper = buildDidHelper(nodeCryptoHelper)
  buildDidRegistryWarpper(didHelper)

  it('creates DID Id', async () => {
    const key = await nodeCryptoHelper.getKey(await nodeCryptoHelper.getRandomBytes(32))
    const id = didHelper.makeDIDId(key)
    const described = id.split(':')
    expect(described.length).toBe(3)
    expect(described).toContain('did')
    expect(described).toContain(process.env.DID_PREFIX)
    expect(typeof described[2]).toBe('string')
  })

  it('creates a DID Document', async () => {
    const key = await nodeCryptoHelper.getKey(await nodeCryptoHelper.getRandomBytes(32))

    const didDocUnsinged = await didHelper.createDID(key, {
      purpose: [DIDPURPOSE_VERIFICATION, DIDPURPOSE_ASSERTION],
      data: 'Hello world!',
      hash: true
    })

    const didDoc = await didHelper.signDID(key, didDocUnsinged)

    testContext['didDoc'] = didDoc
    expect(didDoc).toMatchSnapshot({
      id: expect.any(String),
      verificationMethod: [{
        id: expect.any(String),
        nonce: expect.any(String),
        controller: expect.any(String),
      }],
      assertionMethod: [
        expect.any(String),
      ],
      proof: {
        created: expect.any(String),
        jws: expect.any(String),
        verificationMethod: expect.any(String),
      },
      publicKey: [{
        id: expect.any(String),
        publicKeyBase58: expect.any(String),
      }]
    })
  })

  it('verifies DID Document from itself', async () => {
    if (!testContext.didDoc) {
      throw new Error('No DID doc from pervious test')
    }
    const result = await didHelper.verifyDID(testContext.didDoc)

    expect(result).toBe(true)
  })

  it('produces holder / controller verifiable did', async () => {
    const aliceKey = await nodeCryptoHelper.getKey(await nodeCryptoHelper.getRandomBytes(32))
    aliceKey.nextKeyDigest = 'nkdigest-simulation'
    const bobKey = await nodeCryptoHelper.getKey(await nodeCryptoHelper.getRandomBytes(32))
    bobKey.nextKeyDigest = 'nkdigest-simulation'

    const didDocUnsinged = await didHelper.createDID(aliceKey, {
      purpose: [DIDPURPOSE_VERIFICATION, DIDPURPOSE_ASSERTION],
      data: 'Hello world!',
      hash: true
    })

    const didDoc = await didHelper.signDID(bobKey, didDocUnsinged, VERIFICATION_KEY_CONTROLLER)

    testContext.holderDoc = didDoc
  })

  it('verifies holder veriable did', async () => {
    if (!testContext.holderDoc) {
      throw new Error('No DID doc from pervious test')
    }
    const result = await didHelper.verifyDID(testContext.holderDoc)

    expect(result).toBe(true)
  })

  it('fails on tempered data', async () => {
    if (!testContext.holderDoc) {
      throw new Error('No DID doc from pervious test')
    }

    const brokenDoc = <DIDDocument>JSON.parse(JSON.stringify(testContext.holderDoc))
    if (brokenDoc.verificationMethod && brokenDoc.verificationMethod[0]
      && typeof brokenDoc.verificationMethod[0] === 'object'
      && typeof brokenDoc.verificationMethod[0].proof === 'object') {
      brokenDoc.verificationMethod[0].proof.created = new Date().toUTCString()
    }
    const result = await didHelper.verifyDID(brokenDoc)

    expect(result).toBe(false)
  })
})