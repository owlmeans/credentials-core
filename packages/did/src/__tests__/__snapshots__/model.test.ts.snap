// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DID Helper creates a DID Document 1`] = `
Object {
  "@context": Array [
    "https://w3id.org/did/v1",
    "https://w3id.org/security/v2",
  ],
  "assertionMethod": Array [
    Any<String>,
  ],
  "id": Any<String>,
  "proof": Object {
    "created": Any<String>,
    "jws": Any<String>,
    "proofPurpose": "verificationMethod",
    "type": "EcdsaSecp256k1Signature2019",
    "verificationMethod": Any<String>,
  },
  "publicKey": Array [
    Object {
      "id": Any<String>,
      "publicKeyBase58": Any<String>,
      "type": "EcdsaSecp256k1VerificationKey2019",
    },
  ],
  "verificationMethod": Array [
    Object {
      "controller": Any<String>,
      "id": Any<String>,
      "nonce": Any<String>,
      "type": "EcdsaSecp256k1VerificationKey2019",
    },
  ],
}
`;
